src/puppetlabs/puppetdb/threadpool.clj

```diff
@@ -0,0 +1,119 @@
+           [org.apache.commons.lang3.concurrent BasicThreadFactory BasicThreadFactory$Builder])
+  (:require [clojure.core.async :as async]
+            [clojure.tools.logging :as log]))
+
+(def logging-exception-handler
+  "Exception handler that ensures any uncaught exception that occurs
+  on this thread is logged"
+  (reify Thread$UncaughtExceptionHandler
+    (uncaughtException [_ thread throwable]
+      (log/error throwable "Error processing command"))))
+
+(defn thread-factory
+  "Creates a command thread factory, wrapping the
+  `pool-thread-factory`. Using this thread factory ensures that we
+  have distinguishable names for threads in this pool and that all
+  uncaught exceptions are logged as errors. Without this uncaught
+  exceptions get the default which is standard error"
+  [pool-thread-factory]
+  (-> (BasicThreadFactory$Builder.)
+      (.wrappedFactory pool-thread-factory)
+      (.namingPattern "cmd-proc-thread-%d")
+      (.uncaughtExceptionHandler logging-exception-handler)
+      (.daemon true)
+      .build))
+
+(defrecord GatedThreadpool [^Semaphore semaphore ^ExecutorService threadpool])
+
+(defn create-threadpool
+  "Creates an unbounded threadpool with the intent that access to the
+  threadpool is bounded by the semaphore. Implicitly the threadpool is
+  bounded by `size`, but since the semaphore is handling that aspect,
+  it's more efficient to use an unbounded pool and not duplicate the
+  constraint in both the semaphore and the threadpool"
+  [size]
+  (let [threadpool (ThreadPoolExecutor. 1
+                                        Integer/MAX_VALUE
+                                        1
+                                        TimeUnit/MINUTES
+                                        (SynchronousQueue.))]
+    (->> threadpool
+         .getThreadFactory
+         thread-factory
+         (.setThreadFactory threadpool))
+
+    (map->GatedThreadpool
+     {:semaphore (Semaphore. size)
+      :threadpool threadpool})))
+
+(defn shutdown
+  "Shuts down the gated threadpool, ensuring in-flight work is
+  complete before tearing it down. Currently waits up to 5 minutes for
+  that work to complete"
+  [{:keys [^Semaphore semaphore ^ExecutorService threadpool]}]
+
+  ;; Prevent new work from being accepted
+  (.drainPermits semaphore)
+
+  ;; Shutdown the threadpool, allowing in-flight work to finish
+  (.shutdown threadpool)
+
+  ;; This will block, waiting for the threadpool to shutdown
+  (when-not (.awaitTermination threadpool 5 java.util.concurrent.TimeUnit/MINUTES)
+
+    (log/warn "Attempted to shutdown threadpool, not stopped after 5 minutes, forcibly shutting it down")
+
+    ;; This will force the shutdown of the threadpool and will
+    ;; not allow current threads to finish
+    (.shutdownNow threadpool)
+    (log/warn "Threadpool forcibly shutdown")))
+
+
+(defn call-on-threadpool
+  "Executes `f` on a gated threadpool. Calls `on-complete` with the
+  results of inoking `f`"
+  ([threadpool f]
```

Comment:
param naming is a confusing here; in this case `threadpool` is your defrecord, but below it's an ExecutorService from inside the defrecord.